src/main/java/de/theit/jenkins/crowd/CrowdConfigurationService.java

```diff
@@ -204,8 +209,22 @@ public CrowdConfigurationService(String url, String applicationName, Secret pass
         this.nestedGroups = pNestedGroups;
         this.useSSO = useSSO;
         this.useCache = useCache;
+        this.useTokenCache = useTokenCache;
         this.cacheSize = cacheSize;
         this.cacheTTL = cacheTTL;
+
+
+        if (cacheSize != null && cacheSize > 0) {
+            this.isGroupMemberCache = new CacheMap<>(cacheSize);
+            this.userFromSSOTokenCache = new CacheMap<>(cacheSize);
+            this.userCache = new CacheMap<>(cacheSize);
+            this.groupCache = new CacheMap<>(cacheSize);
+            this.authoritiesForUserCache = new CacheMap<>(cacheSize);
+            if (useTokenCache) {
```

Comment:
@DuMaM

>If useTokecCache will be unchecked, and then Jenkins will reboot, we have no possibility to enable it again without a restart. Line 942 will crash in this situation.

Can you please clarify what you mean? Line 942 is `setValueToCache()` which should be fine as there is a null check prior with a return:
`LOG.log(Level.FINEST, "setValueToCache::cacheObj: {0}", cacheObj.toString());`

Also I tried toggling `useTokenCache` seems to work fine.

Reply:
Line 223 is initializing new cache map (otherwise it's null).
The Constructor for`CrowdConfigurationService` is called only once when the plugin is loading itself.
https://github.com/jenkinsci/crowd2-plugin/blob/436a663dc1c6256b2f7bdc3e97466f58e27193a6/src/main/java/de/theit/jenkins/crowd/CrowdUserDetailsService.java#L69

So, I would rather do something like that:
```          
            this.userFromSSOTokenCache = new CacheMap<>(cacheSize);
            this.userCache = new CacheMap<>(cacheSize);
            this.groupCache = new CacheMap<>(cacheSize);
            this.authoritiesForUserCache = new CacheMap<>(cacheSize);
            this.validationCache = new CacheMap<>(cacheSize);
```

Instead of:
```
            if (useTokenCache) {
                   this.validationCache = new CacheMap<>(cacheSize);
           }
```

Because if somebody uncheck `useTokenCache` and reboot service Jenkins, the plugin will load with `this.validationCache==null`, because `useTokenCache` will be false.
            
I didn't play for long with this code so i may be wrong here.

Reply:
That is correct, the thing is if we initialize the `CacheMap` as you suggested, that in effect turns on the token cache regardless of our UI configuration, the only thing that keeps it from working is the object being null.

Because we will eventually hit `validateSSOAuthentication()` at line 595 and `getValidValueFromCache()` will return null at line 915 if the `CacheMap` has not been initialized.

Therefore this is by design.